### PR TITLE
Raise BetamaxError on unknown and invalid options.

### LINKED ISSUE #117: Fail on invalid matchers?
`valididate_matchers` silently filters out matchers that do not exist. I'm guessing this is to be compatible with vcr. Please correct me if that assumption is wrong.

Ref: https://github.com/sigmavirus24/betamax/blob/51721667484bf25e581cc1435e7085dc53c1aad4/betamax/options.py#L8

I personally would like to hard-fail if a matcher is not found as in my recent testing night there were a number of times I forgot to register the matcher, and was unaware of why my updates to the matcher weren't working.

It would seem that I could manually set `match_options` on the cassette, resulting in an IndexError if a matcher does not exist. However, that doesn't seem like the accepted way to do things.

How would you feel about adding an option, or making it the default, to hard fail when a matcher doesn't exist? I think making the default to fail is good for users, but I understand if that's not desired due to incompatibility with VCR.

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/36824878-fail-on-invalid-matchers?utm_campaign=plugin&utm_content=tracker%2F198445&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F198445&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by bboe**
Resolves #117.

### CODE REVIEW

**Comment by sigmavirus24** [comment]
I'm concerned that using just `BetamaxError` will be not very user friendly. So I think I want us to set up an exception hierarchy here:

- `ValidationError` would inherit from `BetamaxError`
  - Then you could have `InvalidOption` inherit from `ValidationError` so we raise this if the key is not a valid configuration option (L88 in your diff of `betamax/options.py`)
  - And a dictionary of exceptions that map from key to exception, e.g., `match_requests_on` to `MatchersValidationError`. We then get and raise an exception specific to what we're validating.

How does that sound to you?

File: betamax/options.py [hunk 1]
```diff
@@ -84,8 +86,9 @@ def items(self):
     def validate(self):
         for key, value in list(self.data.items()):
             if key not in Options.valid_options:
-                del self[key]
+                raise InvalidOption('{0} is not a valid option'.format(key))
             else:
                 is_valid = Options.valid_options[key]
                 if not is_valid(value):
-                    del self[key]
+                    raise validation_error_map[key]('{0} is not valid'
```
Comment (sigmavirus24) [inline betamax/options.py]: Should we format in the value instead of the key here? Perhaps using `{0!r}` so it's `repr` formatted?

**Comment by bboe** [comment]
@sigmavirus24 Hierarchy added. Doing a few force pushes to fix those few pesky py2.6 assertion method not found errors.

**Comment by bboe** [comment]
Looks like the py2.6 issues have been resolved, and none of the travis builds have failed yet, so I don't see any reason why it would on the next.

Thanks for the feedback!

File: betamax/options.py (hunk 1 above)
Comment (bboe) [inline betamax/options.py]: Resolved (I don't think you get notified for the "👍").

File: tests/unit/test_cassette.py [hunk 2]
```diff
@@ -310,6 +310,12 @@ def test_find_match(self):
         assert i is not None
         assert self.interaction is i
 
+    def test_find_match__missing_matcher(self):
+        self.cassette.match_options = set(['uri', 'method', 'invalid'])
+        self.cassette.record_mode = 'none'
+        self.assertRaises(KeyError, self.cassette.find_match,
```
Comment (sigmavirus24) [inline tests/unit/test_cassette.py]: So we're using py.test. Can we use pytest's check for raises instead, e.g.,

```py
with pytest.raises(KeyError):
    self.cassette.find_match(self.response.request)
```

File: tests/unit/test_exceptions.py [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+def exception_classes():
+    for _, module_object in inspect.getmembers(exceptions):
+        if inspect.isclass(module_object):
+            yield module_object
+
+
+class TestExceptions(unittest.TestCase):
+    def test_all_exceptions_are_betamax_errors(self):
+        for exception_class in exception_classes():
+            self.assertTrue(isinstance(exception_class('msg'),
```
Comment (sigmavirus24) [inline tests/unit/test_exceptions.py]: Same here. If you look at the rest of the test suite, we prefer:

```py
assert isinstance(exception_class('msg'), exceptions.BetamaxError) is True
```

File: tests/unit/test_exceptions.py [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+                                       exceptions.BetamaxError))
+
+    def test_all_validation_errors_are_in_validation_error_map(self):
+        validation_error_map_values = exceptions.validation_error_map.values()
+        for exception_class in exception_classes():
+            if exception_class.__name__ == 'ValidationError' or \
+               not exception_class.__name__.endswith('ValidationError'):
+                continue
+            self.assertTrue(exception_class in validation_error_map_values)
```
Comment (sigmavirus24) [inline tests/unit/test_exceptions.py]: So unittest has `assertIn` but we're using py.test so we should use

```py
assert exception_class in validation_error_map_values
```

File: tests/unit/test_options.py [hunk 5]
```diff
@@ -31,21 +32,46 @@ def test_data_is_valid(self):
         for key in self.data:
             assert key in self.options
 
-    def test_invalid_data_is_removed(self):
+    def test_raise_on_unknown_option(self):
         data = self.data.copy()
         data['fake'] = 'value'
-        options = Options(data)
+        self.assertRaises(exceptions.InvalidOption, Options, data)
```
Comment (sigmavirus24) [inline tests/unit/test_options.py]: Same thing in this file too please

**Comment by bboe** [comment]
@sigmavirus24 thanks for the feedback. I will address the method by which those assertions happen this weekend, as I will be away from the keyboard until then, and unfortunately don't have the five minutes needed to fix it right now.

**Comment by sigmavirus24** [comment]
@bboe don't worry about it then. I'll take care of them tonight or tomorrow morning.

**Comment by bboe** [comment]
Thanks!

**Pull request merged**